scri/asymptotic_bondi_data/bms_charges.py

```diff
@@ -0,0 +1,200 @@
+# Copyright (c) 2020, Michael Boyle
+# See LICENSE file for details: <https://github.com/moble/scri/blob/master/LICENSE>
+
+### NOTE: The functions in this file are intended purely for inclusion in the AsymptoticBondData
+### class.  In particular, they assume that the first argument, `self` is an instance of
+### AsymptoticBondData.  They should probably not be used outside of that class.
+
+import numpy as np
+from math import sqrt, pi
+
+
+def mass_aspect(self, truncate_ell=max):
+    """Compute the Bondi mass aspect of the AsymptoticBondiData.
+
+    The Bondi mass aspect is given by
+
+        M = -Re(\\psi_2 + \\sigma * \\dot{\\bar{\\sigma}})
+
+    Note that the last term is a product between two fields.  If, for example, these both have
+    ell_max=8, then their full product would have ell_max=16, meaning that we would go from
+    tracking 81 modes to 289.  This shows that deciding how to truncate the output ell is
+    important, which is why this function has the extra argument that it does.
+
+    Parameters
+    ==========
+    truncate_ell: int, or callable [defaults to `max`]
+        Determines how the ell_max value of the output is determined.  If an integer is passed,
+        each term in the output is truncated to have at most that ell_max.  (In particular,
+        terms that will not be used in the output are simply not computed, without incurring any
+        errors due to aliasing.)  If a callable is passed, it is passed on to the
+        spherical_functions.Modes.multiply method.  See that function's docstring for details.
+        The default behavior will result in the output having ell_max equal to the largest of
+        any of the individual Modes objects in the equation for M above -- but not the
+        product.
+
+    """
+    if callable(truncate_ell):
+        return -(self.psi2 + self.sigma.multiply(self.sigma.bar.dot, truncator=truncate_ell)).real
+    elif truncate_ell:
+        return -(
+            self.psi2.truncate_ell(truncate_ell)
+            + self.sigma.multiply(self.sigma.bar.dot, truncator=lambda tup: truncate_ell)
+        ).real
+    else:
+        return -(self.psi2 + self.sigma * self.sigma.bar.dot).real
+
+
+def charge_vector_from_aspect(charge):
+    """Output the ell<=1 modes of a BMS charge aspect as the charge four-vector."""
+    four_vector = np.empty(charge.shape, dtype=float)
+    four_vector[..., 0] = charge[..., 0].real
+    four_vector[..., 1] = (charge[..., 1] - charge[..., 3]).real / sqrt(6)
+    four_vector[..., 2] = (charge[..., 1] + charge[..., 3]).imag / sqrt(6)
+    four_vector[..., 3] = charge[..., 2].real / sqrt(3)
+    return four_vector / np.sqrt(4 * np.pi)
```

Comment:
This is bascially the same function that you had for computing the four-momentum vector in the old code in `scri/asymptotic_bondi_data/__init__.py`. Here, we just moved the `1/np.sqrt(4*pi)` into this function instead of in the definition of the charge. We do the same for the rest of the charges too. 

I think the issue here is that our terminology for boost, vector, charge, and apsect is confusing. As a point of reference for discussion, check out the charge equation at Eq. (8) in Dray 1985 (https://iopscience.iop.org/article/10.1088/0264-9381/2/1/002/meta). The function `bondi_boost_charge` here is first computing the integrand of Eq. (8), which we are calling "charge_aspect". Then we use `charge_vector_from_aspect` to integrate the aspect with the generator AND divide by the 1/8π prefactor. This is probably what accounts for the difference between this function and `sf.vector_from_ell_1_modes`. 

There is a lot of juggling of factors, but I remember that @keefemitman, @neevkhera, and I managed to work it all out independently to get this function.

Reply:
The factor of 3 is an entirely sensible (2*ell+1) because these two functions are just doing different things.  I was also toying with mentioning the issue of conformal weight, but I feel like that's too complicated to explain well in a docstring, and is something people using this function should be able to figure out if they need to.  So if we can just add a little explanation to the docstring (like the above) about what the function actually does, I'll be happy.